app/models/host.rb

```diff
@@ -22,7 +22,43 @@ class Host < ApplicationRecord
   validates :name, presence: true
   validates :account, presence: true
 
+  def in_inventory
+    ::HostInventoryAPI.new(
+      User.current.account,
+      ::Settings.host_inventory_url,
+      nil
+    ).inventory_host(id)
+  end
+
   class << self
+    def host_in_inventory(host_id)
+      new(id: host_id).in_inventory
+    end
+
+    def find_or_create_hosts_by_inventory_ids(ids)
+      ids.map { |id| find_or_create_from_inventory(id) }
+    end
+
+    def find_or_create_from_inventory(host_id)
+      existing_hosts = ::Pundit.policy_scope(User.current, self)
+                               .where(id: [host_id])
+
+      existing_hosts.first || create_from_inventory(host_id)
+    end
+
+    def create_from_inventory(host_id)
+      i_host = host_in_inventory(host_id)
+
+      host = find_or_initialize_by(
+        id: i_host['id'],
+        account_id: User.current.account.id
+      ) do |h|
+        h.name = i_host['display_name']
+      end
```

Comment:
`find_or_initialize_by` is something I see as a bit of a red flag, performance-wise. This method, `create_from_inventory` should just create blindly, without checking if the host already exists, and then pass in only new IDs by using a single `where` query to determine what already exists.

Also, it should arguably be `new_from_inventory` so that we can use `Host.import!` to save all records in one query.

I doubt we will find these performance issues during our own testing, but some users may try to associate >1000 systems to a profile, and that's where we'll see timeouts or similar. For this deployment, I'm okay to forgo performance improvements here - we can create a jira to revisit it.